core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -17,25 +17,37 @@
 
 package kafka.server
 
+import kafka.server.AddPartitionsToTxnManager.{verificationFailureRateMetricName, verificationTimeMsMetricName}
 import kafka.utils.Logging
 import org.apache.kafka.clients.{ClientResponse, NetworkClient, RequestCompletionHandler}
 import org.apache.kafka.common.{Node, TopicPartition}
 import org.apache.kafka.common.message.AddPartitionsToTxnRequestData.{AddPartitionsToTxnTransaction, AddPartitionsToTxnTransactionCollection}
 import org.apache.kafka.common.protocol.Errors
 import org.apache.kafka.common.requests.{AddPartitionsToTxnRequest, AddPartitionsToTxnResponse}
 import org.apache.kafka.common.utils.Time
+import org.apache.kafka.server.metrics.KafkaMetricsGroup
 import org.apache.kafka.server.util.{InterBrokerSendThread, RequestAndCompletionHandler}
 
 import java.util
+import java.util.concurrent.TimeUnit
 import scala.collection.mutable
 
 object AddPartitionsToTxnManager {
   type AppendCallback = Map[TopicPartition, Errors] => Unit
+
+  val verificationFailureRateMetricName = "VerificationFailureRate"
```

Comment:
Do we want to add a tag for error code as well? That will help us get per error code failure.

Reply:
Error code makes sense to me.

Reply:
Oh hmm -- so we would have to predefine or lazily create the meters with the various error messages. I'm wondering if this is still necessary. We should be able to figure out the errors from either the api error metrics or from the logging.